@ondrejmirtes is there a way to start working on `phpstan/phpstan-symfony` extension before this is merged and released? I tried to change Composer requirements from `phpstan/phpstan` to `phpstan/phpstan-src` (by adding my fork to `repositories` and using `dev-codito/result-cache-meta-extension as 2.2.0` constraint), I've run `composer -d vendor/phpstan/phpstan-src install` but then I get:

```shell
$ make phpstan

php vendor/bin/phpstan analyse -l 8 -c phpstan.neon src tests
Fatal error: Cannot declare class JetBrains\PHPStormStub\PhpStormStubsMap, because the name is already in use in /.../phpstan/phpstan-symfony/vendor/jetbrains/phpstorm-stubs/PhpStormStubsMap.php on line 10
make: *** [phpstan] Error 255
```

Is it possible to use development version of PHPStan for developing extension, or do we need to finish this first?

--- a/src/Analyser/ResultCache/ResultCacheManager.php
+++ b/src/Analyser/ResultCache/ResultCacheManager.php
@@ -906,7 +906,7 @@ return [
 		return [
 			'cacheVersion' => self::CACHE_VERSION,
 			'phpstanVersion' => ComposerHelper::getPhpStanVersion(),
-			'phpstanExtensions' => $this->getMetaFromPhpStanExtensions(),
+			'metaExtensions' => $this->getMetaFromPhpStanExtensions(),
 			'phpVersion' => PHP_VERSION_ID,
 			'projectConfig' => $projectConfigArray,
 			'analysedPaths' => $this->analysedPaths,

--- a/src/Analyser/ResultCache/ResultCacheManager.php
+++ b/src/Analyser/ResultCache/ResultCacheManager.php
@@ -10,6 +10,7 @@ use PHPStan\Collectors\CollectedData;
 use PHPStan\Command\Output;
 use PHPStan\Dependency\ExportedNodeFetcher;
 use PHPStan\Dependency\RootExportedNode;
+use PHPStan\DependencyInjection\Container;
 use PHPStan\DependencyInjection\ProjectConfigHelper;
 use PHPStan\File\CouldNotReadFileException;
 use PHPStan\File\FileFinder;
@@ -66,6 +67,7 @@ final class ResultCacheManager
 	 * @param string[] $scanDirectories
 	 */
 	public function __construct(
+		private Container $container,
 		private ExportedNodeFetcher $exportedNodeFetcher,
 		private FileFinder $scanFileFinder,
 		private ReflectionProvider $reflectionProvider,
@@ -315,6 +317,11 @@ final class ResultCacheManager
 			ksort($currentMeta['projectConfig']);
 
 			$currentMeta['projectConfig'] = Neon::encode($currentMeta['projectConfig']);
+		}
+
+		// Do not invalidate result cache generated before introducing support for result cache meta extensions
+		if (!isset($currentMeta['phpstanExtensions'])) {
+			$currentMeta['phpstanExtensions'] = [];
 		}
 
 		return $cachedMeta !== $currentMeta;
@@ -904,6 +911,7 @@ return [
 		return [
 			'cacheVersion' => self::CACHE_VERSION,
 			'phpstanVersion' => ComposerHelper::getPhpStanVersion(),
+			'phpstanExtensions' => $this->getMetaFromPhpStanExtensions(),
 			'phpVersion' => PHP_VERSION_ID,
 			'projectConfig' => $projectConfigArray,
 			'analysedPaths' => $this->analysedPaths,
@@ -1034,6 +1042,29 @@ return [
 		ksort($stubFiles);
 
 		return $stubFiles;
+	}
+
+	/**
+	 * @return array<string, string>
+	 * @throws ShouldNotHappenException
+	 */
+	private function getMetaFromPhpStanExtensions(): array
+	{
+		$meta = [];
+
+		/** @var ResultCacheMetaExtension $extension */
+		foreach ($this->container->getServicesByTag(ResultCacheMetaExtension::EXTENSION_TAG) as $extension) {
+			if (array_key_exists($extension->getKey(), $meta)) {
+				throw new ShouldNotHappenException(sprintf(
+					'Duplicate ResultCacheMetaExtension with key "%s" found.',
+					$extension->getKey(),
+				));
+			}
+
+			$meta[$extension->getKey()] = $extension->getHash();
+		}
+
+		return $meta;
 	}
 
 }

--- a/src/Analyser/ResultCache/ResultCacheManager.php
+++ b/src/Analyser/ResultCache/ResultCacheManager.php
@@ -319,11 +319,6 @@ final class ResultCacheManager
 			$currentMeta['projectConfig'] = Neon::encode($currentMeta['projectConfig']);
 		}
 
-		// Do not invalidate result cache generated before introducing support for result cache meta extensions
-		if (!isset($currentMeta['phpstanExtensions'])) {
-			$currentMeta['phpstanExtensions'] = [];
-		}
-
 		return $cachedMeta !== $currentMeta;
 	}
 

--- a/src/Analyser/ResultCache/ResultCacheMetaExtension.php
+++ b/src/Analyser/ResultCache/ResultCacheMetaExtension.php
@@ -1,0 +1,39 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Analyser\ResultCache;
+
+/**
+ * ResultCacheMetaExtension can be used for extending PHPStan's built-in mechanism that is used for
+ * calculating metadata for result cache. Using this extension you may add additional metadata that will
+ * be used for determining if analysis must be run again, or can be re-used from result cache.
+ *
+ * @see https://github.com/phpstan/phpstan-symfony/issues/255 for the context.
+ *
+ * To register it in the configuration file use the `phpstan.resultCacheMetaExtension` service tag:
+ *
+ * ```
+ * services:
+ * 	-
+ *		class: App\PHPStan\MyExtension
+ *		tags:
+ *			- phpstan.resultCacheMetaExtension
+ * ```
+ *
+ * @api
+ */
+interface ResultCacheMetaExtension
+{
+
+	public const EXTENSION_TAG = 'phpstan.resultCacheMetaExtension';
+
+	/**
+	 * Returns unique key for this result cache meta entry. This describes the source of the metadata.
+	 */
+	public function getKey(): string;
+
+	/**
+	 * Returns hash of the result cache meta entry. This represents the current state of the additional meta source.
+	 */
+	public function getHash(): string;
+
+}

--- a/src/DependencyInjection/ConditionalTagsExtension.php
+++ b/src/DependencyInjection/ConditionalTagsExtension.php
@@ -5,6 +5,7 @@ namespace PHPStan\DependencyInjection;
 use Nette;
 use Nette\DI\CompilerExtension;
 use Nette\Schema\Expect;
+use PHPStan\Analyser\ResultCache\ResultCacheMetaExtension;
 use PHPStan\Analyser\TypeSpecifierFactory;
 use PHPStan\Broker\BrokerFactory;
 use PHPStan\Collectors\RegistryFactory as CollectorRegistryFactory;
@@ -59,6 +60,7 @@ final class ConditionalTagsExtension extends CompilerExtension
 			LazyParameterOutTypeExtensionProvider::METHOD_TAG => $bool,
 			LazyParameterOutTypeExtensionProvider::STATIC_METHOD_TAG => $bool,
 			DiagnoseExtension::EXTENSION_TAG => $bool,
+			ResultCacheMetaExtension::EXTENSION_TAG => $bool,
 		])->min(1));
 	}
 

